src/asr/asr_pytorch.py

```diff
@@ -132,35 +132,49 @@ def update_core(self):
             logging.warning('grad norm is nan. Do not update model.')
         else:
             optimizer.step()
-        delete_feat(x)
-
-
-class DataParallel(torch.nn.DataParallel):
-    def scatter(self, inputs, kwargs, device_ids, dim):
-        r"""Scatter with support for kwargs dictionary"""
-        if len(inputs) == 1:
-            inputs = inputs[0]
-        avg = int(math.ceil(len(inputs) * 1. / len(device_ids)))
-        # inputs = scatter(inputs, device_ids, dim) if inputs else []
-        inputs = [[inputs[i:i + avg]] for i in range(0, len(inputs), avg)]
-        kwargs = torch.nn.scatter(kwargs, device_ids, dim) if kwargs else []
-        if len(inputs) < len(kwargs):
-            inputs.extend([() for _ in range(len(kwargs) - len(inputs))])
-        elif len(kwargs) < len(inputs):
-            kwargs.extend([{} for _ in range(len(inputs) - len(kwargs))])
-        inputs = tuple(inputs)
-        kwargs = tuple(kwargs)
-        return inputs, kwargs
-
-    def forward(self, *inputs, **kwargs):
-        if not self.device_ids:
-            return self.module(*inputs, **kwargs)
-        inputs, kwargs = self.scatter(inputs, kwargs, self.device_ids, self.dim)
-        if len(self.device_ids) == 1:
-            return self.module(*inputs[0], **kwargs[0])
-        replicas = self.replicate(self.module, self.device_ids[:len(inputs)])
-        outputs = self.parallel_apply(replicas, inputs, kwargs)
-        return self.gather(outputs, self.output_device)
+
+
+class CustomConverter(object):
+    """CUSTOM CONVERTER"""
+
+    def __init__(self, device, subsamping_factor=1):
+        self.device = device
+        self.subsamping_factor = subsamping_factor
+        self.ignore_id = -1
+
+    def __call__(self, batch):
+        # batch should be located in list
+        assert len(batch) == 1
+        batch = batch[0]
+
+        # load acoustic features and target sequence of token ids
```

Comment:
We can unify this data preprocessing part (before sending the data to GPU) with the one in `asr_chainer.py` as one function.

Reply:
In addition, it is useful to prefetch if loading from scp file becomes other function. cf. #340